test_util/launch.py

```diff
@@ -61,40 +68,103 @@ def delete(self, info):
         raise NotImplementedError()
 
-    def test(self, info):
-        raise NotImplementedError()
+    def test(self, info, test_cmd):
+        ssh_info = info['ssh']
+        try:
+            check_keys(ssh_info, ['user', 'private_key'])
+        except LauncherError:
+            print('DC/OS Launch is missing sufficient SSH info to run tests!')
+            raise
+        details = self.describe(info)
+        test_host = details['masters'][0]['public_ip']
+        with ssh.tunnel.tunnel(ssh_info['user'], ssh_info['private_key'], test_host) as test_tunnel:
+            return test_util.runner.integration_test(
+                tunnel=test_tunnel,
+                dcos_dns=test_host,
+                master_list=[m['private_ip'] for m in details['masters']],
+                agent_list=[a['private_ip'] for a in details['private_agents']],
+                public_agent_list=[a['private_ip'] for a in details['public_agents']],
+                aws_access_key_id=info['provider'].get('access_key_id'),
+                aws_secret_access_key=info['provider'].get('secret_access_key'),
+                region=info['provider'].get('region'),
+                test_cmd=test_cmd)
 
 
 class AwsCloudformationLauncher(AbstractLauncher):
     def __init__(self, boto_wrapper):
         self.boto_wrapper = boto_wrapper
 
     def create(self, config):
+        """
+        Checks config and creates missing resources as necessary
+        If actual stack creation fails, then network resources
+        will be cleaned up
+
+        Note: ssh_from_config and provide_network_prereqs will mutate
+            config
+        """
         check_keys(config, ['stack_name', 'template_url'])
-        ssh_info = self.ssh_from_config(config)
-        # NOTE: even if parameters not given, ssh_from_config will add KeyName
-        # parameter to config['parameters']
-        self.boto_wrapper.create_stack(
-            config['stack_name'], config['template_url'], config['parameters'])
+        ssh_info, temp_resources = self.ssh_from_config(config)
+        if config.get('zen_helper', False):
+            temp_resources.update(self.provide_network_prereqs(config))
+        try:
+            self.boto_wrapper.create_stack(
+                config['stack_name'], config['template_url'], config['parameters'])
+        except Exception as ex:
+            self.delete_temp_resources(temp_resources)
+            raise LauncherError('ProviderError', None) from ex
         return {
             'type': 'cloudformation',
             'stack_name': config['stack_name'],
             'provider': {
                 'region': config['provider_info']['region'],
                 'access_key_id': config['provider_info']['access_key_id'],
                 'secret_access_key': config['provider_info']['secret_access_key']},
-            'ssh': ssh_info}
+            'ssh': ssh_info,
+            'temp_resources': temp_resources}
+
+    def provide_network_prereqs(self, config):
```

Comment:
4 conditional flags would be pretty nasty looking. Let's talk this one over in person